src/linux/cgroups.cpp

```diff
@@ -1403,9 +1403,15 @@ class TasksKiller : public Process<TasksKiller>
 protected:
   void initialize() override
   {
-    // Stop when no one cares.
-    promise.future().onDiscard(lambda::bind(
-        static_cast<void (*)(const UPID&, bool)>(terminate), self(), true));
+    // We don't want to stop immediately upon discard, because
+    // it could leave the cgroup frozen which means that processes
+    // are stuck in uninterrutible sleep (D state), which is quite bad.
+    // So upon discard we still do our best and keep trying to
+    // kill the cgroup for up to FREEZE_RETRY_INTERVAL which should be
+    // a reasonable upper bound.
+    promise.future().onDiscard([this]() {
+      delay(FREEZE_RETRY_INTERVAL, self(), &Self::selfTerminate);
```

Comment:
Would there be a way to avoid this extra delay if we detect the cgroup is not frozen at that time?

Reply:
Hey @kamaradclimber , thanks for reviewing :).

It's just a delay for the discard: the killer process will continue running, which means that it will do its usual freeze/kill/thaw, and therefore will stop as soon as the tasks have been killed, it won't wait any longer than necessary. Since usually a cgroup can be frozen and killed very quickly - few ms - it should have negligible impact.

Reply:
:ok_hand: